File: Progetto1/Python/solve_linear.py
import numpy
from scipy import io
from scipy.sparse.linalg import spsolve
from scipy.sparse import csc_matrix
import matplotlib.pylab as plt
import numpy as np
from numpy import linalg as LA
import time

# N.B. once a matrix has been constructed, 
# convert to CSR or CSC format for fast arithmetic 
# and matrix vector operations
# consider using the COO format when constructing large matrices

def solve_linear(matrix):
    #Load sparse matrix
    A = io.mmread(matrix)

    #Plot the spy of the matrix
    #plt.spy(A)
    #plt.show()

    #Starting time
    start = time.time()

    #Exact solution
    matrix_size = A.shape
    xe = np.ones((matrix_size[0],1))

    #Compute value of b
    b = A * xe

    #Compress matrix
    A = csc_matrix(A)
    b = csc_matrix(b)

    #Solve linear system
    x = spsolve(A, b, use_umfpack = True)

    #End time
    el_time = time.time() - start 

    #Relative error
    err = LA.norm(numpy.subtract(x, xe)) / LA.norm(xe)

    return err, el_time

File: Progetto1/Matlab/solve_linear.m
%function [x, err, timerVal] = solve_linear(matrix)
    
function [x, err, timerVal, memory_in_mb] = solve_linear(matrix)

    %INPUT:
    %matrix = name of the matrix
    
    %OUPUT:
    %x = solution of the system
    %err = relative error
    %memory_in_mb = memory in mb to solve the linear system

    %Load sparse matrix
    matrix = load(matrix);
    A = matrix.Problem.A;
    %spy(A);
    
    %Exact solution
    matrix_size = size(A);
    xe = ones(matrix_size(1), 1);

    %Compute value of b
    b = A * xe;
    
    %Start calculating time and memory
    tic;
   
    %Solve linear system
    x = A\b;
    
    [user,~] = memory;
    timerVal = toc;

    %1 MB = 1048576 Bytes
    memory_in_mb = user.MemUsedMATLAB / 1048576;
    
    %Relative error
    err = norm(x - xe, 2) / norm(xe, 2);
    
end